### ISSUE TITLE
Dataflow performance issue

### ISSUE THREAD

**Issue by gentryx** [STEllAR-GROUP/hpx#2469]
I've written a performance test ([HPXDataflowGold](https://github.com/gentryx/libgeodecomp/blob/master/src/testbed/hpxperformancetests/main.cpp#L551)) for LibGeoDecomp's [HPXDataflowSimulator](https://github.com/gentryx/libgeodecomp/blob/master/src/parallelization/hpxdataflowsimulator.h) (used for DGSWEM in STORM) and the performance I've measured seems to be off by a huge margin. My single-threaded, not dataflow-based reference code ([HPXDataflowVanilla](https://github.com/gentryx/libgeodecomp/blob/master/src/testbed/hpxperformancetests/main.cpp#L591)) is about 10x faster. So I'm curious if I'm using dataflow wrong or if it is an HPX issue.

The performance test uses a simple [test model](https://github.com/gentryx/libgeodecomp/blob/master/src/testbed/hpxperformancetests/main.cpp#L411) for which a grid of 100x100 cells is instantiated. For each time step every cell will send an object [MessageType](https://github.com/gentryx/libgeodecomp/blob/master/src/testbed/hpxperformancetests/main.cpp#L383) to each of its up to four neighbors (left, right, top, bottom). The cells will also evaluate the messages received from their neighbors to ensure correctness. 1000 time steps are simulated.

My reference code [HPXDataflowVanilla](https://github.com/gentryx/libgeodecomp/blob/master/src/testbed/hpxperformancetests/main.cpp#L591)stores all messages in a `std::vector<std::map<int, MessageType> >` and on my notebook (Intel Core i7-6700HQ, 4 physical cores, 8 logical cores, 2.6 GHz base clock) I get up to 4.01 MLUPS (mega lattice updates per second) on a single core. The HPX dataflow code yields up to 0.28 MLUPS on four cores (and 0.098 MLUPS on a single core).

In other words: the HPX dataflow code spends about 36000 cycles on a single cell update while the reference code only needs 650.

Any help is appreciated, especially WRT optimizing the [HPXDataflowSimulator](https://github.com/gentryx/libgeodecomp/blob/master/src/parallelization/hpxdataflowsimulator.h).

**Comment by hkaiser** [comment]
Unfortunately, I can't build libgeodecomp with MSVC so I can't really look what's going on. My guess would be however, that the amount of work executed in the HPX-threads spawned (either by dataflow or otherwise) is too small to warrant creating a new HPX-thread. The overhead for creating an HPX-thread is in the range of 1-2 microseconds (depending on the context). That means that if the amount of work executed by this thread is very small (and 650 cycles doesn't sound like much work), the overall execution time will be dominated by the overheads introduced by launching new HPX-threads.

In our experience it is necessary to have an average execution time of ~100-120 microseconds per HPX-thread in order to amortize the associated overheads.

**Comment by gentryx** [comment]
As discussed offline I'm closing this as it's just the expected scheduling behavior. Thanks!
